### PR TITLE
refactor: reintroduce the new entry point

### PR DESCRIPTION
**Pull request by barak007**

### CODE REVIEW

File: packages/core/src/runtime-main.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        (globalThis as any).name = instanceId;
+    }
+
+    const buildConfigPromise = runtimeConfiguration.importConfig(configName);
+    const runtimeConfigPromise = runtimeConfiguration.load(env.env, featureName, configName);
+
+    const [buildConfig, runtimeConfig] = await Promise.all([buildConfigPromise, runtimeConfigPromise]);
+
+    // only load features after the config is loaded to avoid blocking onload event
+    const { entryFeature, resolvedContexts } = await featureLoader.loadEntryFeature(featureName, {});
```
Comment (barak007) [inline packages/core/src/runtime-main.ts:51]: this was the pain point

File: packages/runtime-node/src/node-environment.ts [hunk 2]
```diff
@@ -47,22 +47,9 @@ export async function runNodeEnvironment<ENV extends AnyEnvironment>({
...
-    const featureLoader = new FeatureLoadersRegistry(new Map(Object.entries(featureLoaders)), resolvedContexts);
-    const optionsRecord: Record<string, string | boolean> = {};
 
-    for (const [key, val] of options || []) {
-        optionsRecord[key] = val;
-    }
-    const loadedFeatures = await featureLoader.getLoadedFeatures(featureName, optionsRecord);
-    const runningFeatures = [loadedFeatures[loadedFeatures.length - 1]!];
+    const featureLoader = new FeatureLoadersRegistry(new Map(Object.entries(featureLoaders)));
+    const { entryFeature, resolvedContexts } = await featureLoader.loadEntryFeature(featureName, toRecord(options));
```
Comment (daomry) [inline packages/runtime-node/src/node-environment.ts:52]: why not use `Object.fromEntries(options || [])` ?

File: packages/runtime-node/src/node-environment.ts [hunk 3]
```diff
@@ -138,3 +125,11 @@ export function createFeatureLoaders(
     }
     return featureLoaders;
 }
+
+function toRecord(options: [string, string | boolean][] | undefined) {
```
Comment (daomry) [inline packages/runtime-node/src/node-environment.ts:129]: very close to a Object.fromEntries clone, can it be removed?

File: packages/core/src/run-engine-app.ts [hunk 4]
```diff
@@ -78,7 +75,8 @@ export class FeatureLoadersRegistry {
      */
     async getLoadedFeatures(
         rootFeatureName: string,
-        runtimeOptions: Record<string, string | boolean> = {}
+        runtimeOptions: Record<string, string | boolean>,
+        resolvedContexts: Record<string, string>
```
Comment (daomry) [inline packages/core/src/run-engine-app.ts:79]: now that the change is tested, can we keep the old defaults API (rootFeatureName , ...={}, ...={}) ?

File: packages/core/src/runtime-configurations.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+export type ConfigModule = {
+    default: TopLevelConfig;
+};
+
+export type ConfigLoader = () => Promise<ConfigModule[]>;
+
+export type ConfigLoaders = Record<string, ConfigLoader>;
+
+export class RuntimeConfigurations {
```
Comment (daomry) [inline packages/core/src/runtime-configurations.ts:17]: plz add comment explain what this class is used for an how... and read what gpt writes

**Review by daomry — APPROVED** [review]

**Review by daomry — CHANGES_REQUESTED** [review]
plz add ts-doc comments for exported classes/functions

File: packages/runtime-node/src/node-environment.ts (hunk 2 above)
Comment (barak007) [reply packages/runtime-node/src/node-environment.ts]: done

File: packages/runtime-node/src/node-environment.ts (hunk 3 above)
Comment (barak007) [reply packages/runtime-node/src/node-environment.ts]: removed

File: packages/core/src/runtime-configurations.ts (hunk 5 above)
Comment (barak007) [reply packages/core/src/runtime-configurations.ts:20]: added comments to each public function and the class itself

**Review by daomry — APPROVED** [review]

**Pull request merged**